content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,792 @@
+  ret void
+}
+
+define internal void @simple(<...>) (.omp_outlined_debug__) <...> {
+<...>
+  call void @__kmpc_for_static_init_4(ptr @1, i32 %8, i32 34, <...>)
+<...>
+omp.inner.for.body:
+<...>
+omp.precond.end:
+  ret void
+}
+```
+
+`simple` no longer does the work of the program. Instead it uses
+`__kmpc_fork_call` to call another version of `simple`, which calls yet another
+version. `__kmpc_fork_call` is part of the OpenMP
+[runtime library](https://github.com/llvm/llvm-project/tree/a23260087db14032094d62dcf8be9be6f4e3d788/openmp). The final
+`.omp_outlined_debug__` version of `simple` is where the work actually happens.
+
+It includes labels like `omp.inner.for.body` but these are not specific to OpenMP.
+They are just normal LLVM IR labels.
+
+## Bringing OpenMP to MLIR
+
+Early in 2020, Kiran Chandramohan (Arm) [proposed](https://discourse.llvm.org/t/rfc-openmp-dialect-in-mlir/397) an MLIR dialect for OpenMP, with the end goal
+of Flang supporting OpenMP.
+
+> We (Arm) started the work for the OpenMP MLIR dialect because of Flang.
+> ... So, MLIR has an OpenMP dialect because of Flang.
+* Kiran Chandramohan
+
+This dialect would represent OpenMP specifically, unlike the generic LLVM IR you
+get from Clang.
+
+If you go back to the Fortran OpenMP example and compile it without OpenMP
+enabled, you get this MLIR:
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPsimple(<...> {
+    %1:2 = hlfir.declare %arg0 <...> {uniq_name = "_QFsimpleEn"} : <...>
+    %3:2 = hlfir.declare %2 <...> {uniq_name = "_QFsimpleEi"} : <...>
+    %10:2 = hlfir.declare %arg1(%9) <...> {uniq_name = "_QFsimpleEa"} : <...>
+    %17:2 = hlfir.declare %arg2(%16) <...> {uniq_name = "_QFsimpleEb"} : <...>
+    %22:2 = fir.do_loop <...> {
+      <...>
+      hlfir.assign %34 to %37 : f32, !fir.ref<f32>
+    }
+    fir.store %22#1 to %3#1 : !fir.ref<i32>
+    return
+  }
+}
+```
+(output heavily edited for readability)
+
+Notice that the `DO` loop has been converted into `for.do_loop`. Now enable
+OpenMP and compile again:
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPsimple(<...> {
+    %1:2 = hlfir.declare %arg0 <...> {uniq_name = "_QFsimpleEn"} : <...>
+    %10:2 = hlfir.declare %arg1(%9) <...> {uniq_name = "_QFsimpleEa"} : <...>
+    %17:2 = hlfir.declare %arg2(%16) <...> {uniq_name = "_QFsimpleEb"} : <...>
+    omp.parallel {
+      %19:2 = hlfir.declare %18 {uniq_name = "_QFsimpleEi"} : <...>
+      omp.wsloop {
+        omp.loop_nest (%arg3) : i32 = (%c2_i32) to (%20) inclusive step (%c1_i32) {
+          hlfir.assign %32 to %35 : f32, !fir.ref<f32>
+          omp.yield
+        }
+      }
+      omp.terminator
+    }
+    return
+  }
+}
+```
+(output heavily edited for readability)
+
+You will see that instead of `for.do_loop` you have `omp.parallel`. `omp` is
+MLIR's [OpenMP dialect](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/).
+This is a higher level and much more literal translation of the
+`parallel` directive than Clang produced.
```

Comment:
```suggestion
`parallel do` combined directive than Clang produced.
```

Reply:
directive / combined directive, it's the same thing to the casual reader in my opinion. So I will leave it as is unless it's misleading?

I have refactored this bit some so the wording is not exactly the same. I tried to make it clear that Clang and Flang aren't processing literally the same directive here, but the intent is the same.

Reply:
Right, the distinction between directive / combined directive should not matter to the casual reader.

I saw that in the latest version of the text you have used `parallel do` and `parallel for` instead, so I think it's fine now.